File: src/main/python/Day11.py
import Utils as Utils

def spread(lines, visited, x, y, width, height):
    index = x + y * width
    if index in visited:
        return
    visited.add(index)
    
    for i in [ 0, 1, 2, 3, 5, 6, 7, 8 ]:
        nx = x + (i % 3) - 1
        ny = y + (i // 3) - 1
        if nx >= 0 and nx < width and ny >= 0 and ny < height:
            lines[nx + ny * width] += 1
            if lines[nx + ny * width] > 9:
                spread(lines, visited, nx, ny, width, height)

def performSteps(lines, width, height):
    visited = set()
    
    # Add one to all elements in the array
    for i in range(width * height):
        lines[i] += 1
    
    # Then for all values that are greater than nine propagate the spread
    for i in range(width * height):
        if lines[i] > 9:
            spread(lines, visited, i % width, i // width, width, height)
    
    # All octopus that has flasheed will be set to zero
    for i in visited:
        lines[i] = 0
    
    return len(visited)

@Utils.time_it_and_evaluate
def partOne(lines, width, height):
    score = 0
    for i in range(100):
        score += performSteps(lines, width, height)
    return score

@Utils.time_it_and_evaluate
def partTwo(lines, width, height):
    index = 0
    while True:
        score = performSteps(lines, width, height)
        index += 1
        
        if score == len(lines):
            return index

if __name__ == '__main__':
    lines = Utils.readAllLines(2021, 'day11')
    values = [ int(x) for x in ''.join(lines) ]
    width = len(lines[0])
    height = len(lines)
    
    print('Day 11')
    partOne(values.copy(), width, height)
    partTwo(values.copy(), width, height)

File: src/main/java/hardcoded/aoc2021/Day11.java
package hardcoded.aoc2021;

import java.util.*;

import hardcoded.util.Utils;

public class Day11 {
	public static void main(String[] args) throws Exception {
		List<String> lines = Utils.readAllLines(2021, "day11");
		int width = lines.get(0).length();
		int height = lines.size();
		
		int[] array = String.join("", lines).chars().map(i -> (i - '0')).toArray();
		
		Utils.printf("Day 11\n");
		Utils.printf("PartOne: %d\n", partOne(array.clone(), width, height));
		Utils.printf("PartTwo: %d\n", partTwo(array.clone(), width, height));
	}
	
	private static void spread(int[] array, Set<Integer> visited, int x, int y, int width, int height) {
		int index = x + y * width;
		if(!visited.add(index)) return;
		
		for(int i = 0; i < 9; i++) {
			int dy = (i / 3) - 1;
			int dx = (i % 3) - 1;
			if(dy != 0 || dx != 0) {
				int nx = x + dx;
				int ny = y + dy;
				
				if(nx >= 0 && nx < width && ny >= 0 && ny < height && ++array[nx + ny * width] > 9) {
					spread(array, visited, nx, ny, width, height);
				}
			}
		}
	}
	
	private static int performSteps(int[] array, int width, int height) {
		Set<Integer> visited = new HashSet<>();
		
		// First add one to all elements in the array.
		for(int i = 0, len = array.length; i < len; i++) {
			array[i]++;
		}
		
		// Then for all values that are greater than nine propagate the spread.
		for(int i = 0, len = array.length; i < len; i++) {
			if(array[i] > 9) {
				spread(array, visited, i % width, i / width, width, height);
			}
		}
		
		// All octopus that has flashed will be set to zero.
		for(int i : visited) {
			array[i] = 0;
		}
		
		return visited.size();
	}
	
	public static long partOne(int[] array, int width, int height) throws Exception {
		int score = 0;
		for(int i = 0; i < 100; i++) {
			score += performSteps(array, width, height);
		}
		return score;
	}
	
	public static long partTwo(int[] array, int width, int height) throws Exception {
		int index = 0;
		while(true) {
			int score = performSteps(array, width, height);
			index++;
			
			// If the whole board flashes it will have a score of the length of the array.
			if(score == array.length) {
				return index;
			}
		}
	}
}
